### PR TITLE
Added a function to convert exif dates and times to datetime objects to enable datetime fields on exif data

### LINKED ISSUE #266: Enable datetime fields on exif data
It is difficult to change filenames based on exif data.  The provided options of using {created} and {lastmodified} filters are not useful when the incorrect FileSystem dates do not match the correct exif data dates

**Describe the solution you'd like**
The ability to use datetime objects on exif timestamps, e.g., {exif.image.datetime.month} and {exif.image.datetime.year} 

**Describe alternatives you've considered**

The only thing I can think of is to use exiftool to modify the file first, then run organize tool , but it shouldn't be this difficult.
For exiftool, I ran `exiftool -time:all -G1 -a -s FILENAME.HEIC` to see what datetimes were available.  The correct datetime for mine happened to be `DateCreated` .
To modify the file i ran `exiftool "-FileModifyDate<DateCreated" FILENAME.HEIC`
This changed the modify date AND the FileCreateDate to match the `DateCreated` exif data.   I could then use the `{created.month}` as well as the `{lastmodified.month} `fields with the organize tool .

Would have been much nicer to just use the `{exif.image.datetime.month}`

### PR DESCRIPTION
**Pull request by FlorianFritz**
## Summary

Added a function to convert exif dates and times to datetime objects by following rules:
- If `key` contains "datetime" convert `value` to 'datetime.datetime'
        (e.g. image.datetime, exif.datetimeoriginal, exif.datetimedigitized)
- If `key` contains "date" convert `value` to 'datetime.date'
        (e.g. gps.date)
- If `key` contains "offsettime" convert `value` to 'datetime.timezone'
        (e.g. exif.offsettimeoriginal, exif.offsettimedigitized)
- Otherwise `value` is not converted and returned as is

## Related issue number

fix #266 : This PR closes feature request [Enable datetime fields on exif data](https://github.com/tfeldmann/organize/issues/266)

## Checklist

- [x] Tests for the changes exist and pass on my machine and [CI](https://github.com/FlorianFritz/organize/actions/runs/5414671302)*
- [x] Documentation reflects the changes where applicable
- [x] Added change to CHANGELOG.md
- [x] My PR is ready to review

*) Some CI-Actions failed on 'Install dependencies', the others succeeded.

### CODE REVIEW

**Comment by temirov** [comment]
That seems like a great addition to the tool. Is there something missing in the PR, @tfeldmann ?

**Comment by jleatham** [comment]
@FlorianFritz , @tfeldmann ,
TL;DR: {exif.image.datetime.month} is still not working

Regarding issue 266: https://github.com/tfeldmann/organize/issues/266
Regarding PR 284: https://github.com/tfeldmann/organize/pull/284

Thanks for the response, however, I am not sure how to use this?
I've updated to 2.4.2, but when I try and use exif data in the rules I still get errors.  

The goal is to use the exif datetime month so that I can ignore the incorrect file datetime and read the correct exif datetime so that I can rename the photos and put them in their respective folders:
```
... # what I use today that gives incorrect file based datetime
      - move: 
          dest: "/Volumes/Memories/photos_to_sift/test/{created.year}/{created.month}/{created.strftime('%Y-%m-%d--%H-%M-%S')}-{name}.{extension}"
```

```
... # what I would like to be able to use
      - move: 
          dest: "/Volumes/Memories/photos_to_sift/test/{exif.image.year}/{exif.image.month}/{exif.image.strftime('%Y-%m-%d--%H-%M-%S')}-{name}.{extension}"
```


Recreate problem:
```
... # use exif filter , and a location that contains images with exif data
    actions:
      - echo: "exif datetime: {exif.image.datetime}"           # Works but is not useful
      - echo: "exif month: {exif.image.datetime.month}"       # Fails
      - echo: "exif month: {exif.image.month}"                   # Also Fails
 ...
```

**Comment by FlorianFritz** [comment]
Good Morning @jleatham,

have a look at `test_exif_filter_datetime` in [tests/filters/test_exif.py#L78](https://github.com/tfeldmann/organize/blob/10ca44fb360df4041d09bd88aaa723291be07898/tests/filters/test_exif.py#L78) for the correct syntax: 

You have to add 
```yaml 
- exif:
  exif.<fieldname>
```
to your `filters` section first and use the values as follows:
`{exif.exif.<fieldname>.<datetime_function>()}`
e.g. `{exif.exif.datetimeoriginal.isoformat()}`

If it doesn’t work, get back to me and I‘ll try to write an example.

Best regards 
Florian

**Comment by jleatham** [comment]
I couldn't seem to get this to work, I apologize.  Thank you for your help.
Here is what my exif data looks like when I run this rule:
```
  - name: "Organize pictures"
    locations:
      - /Volumes/Memories/photos_to_sift/
    subfolders: true
    filters:
      - extension: 
        - heic
      - created
      - name
      - exif
      - lastmodified
    actions:
      - echo: "exif datetime: {exif.image.datetime}"
      - echo: "{exif}"
      - echo: "Create date: {created}"
      - move: 
          dest: "/Volumes/Memories/photos_to_sift/test/{created.year}/{created.month}/{created.strftime('%Y-%m-%d--%H-%M-%S')}-{name}.{extension}"
          on_conflict: skip
```

Result (sim):
```
    - (echo) exif datetime: 2023:06:17 12:46:30
    - (echo) {'image': {... 'datetime': '2023:06:17 12:46:30', ...}, 'gps':
{...}, 'exif': {... 'datetimeoriginal': '2023:06:17 12:46:30', ...}}
    - (echo) Create date: 2023-06-21 20:32:20-05:00
    - (move) Move to /Volumes/Memories/photos_to_sift/test/2023/6/2023-06-21--20-32-20-IMG_4307.HEIC
```


So now, when I try and read the exif data using the new filter, I receive the error: `(echo) ERROR! 'str object' has no attribute 'isoformat'`

The new rule:
```
...
      - exif:
          exif.datetimeoriginal
...
    actions:
      - echo: "exif datetime: {exif.exif.datetimeoriginal.isoformat()}"
...
```

What am I doing wrong?  And furthermore, is there a way to access the datetime for naming purposes similar to {created.month} ?  I tried to use exif.month by {exif.exif.datetimeoriginal.month} ., but that just comes up blank in the echo.
@FlorianFritz

**Comment by FlorianFritz** [comment]
Hi @jleatham,

I've doublechecked your problem: You have to have the latest git-version installed:

```bash
# Remove old version (<=2.4.2)
$ pip uninstall organize-tool
# Install git version
$ pip install git+https://github.com/tfeldmann/organize.git
```

To check if you have the new code:
```bash
$ less /<path-to-site-packages>/organize/filters/exif.py
e.g.
less /opt/miniconda3/lib/python3.10/site-packages/organize/filters/exif.py
```

In line 14 and following you should see the `to_datetime` function:
```python
def to_datetime(key: str, value: str) -> ExifValue:
    """Converts exif datetime/date/offsettime fields to datetime objects
    ...
```

Best regards
Florian

**Comment by jleatham** [comment]
Thanks! That now works for HEIC files with that particular exif format.  

Notes: I had to run `xcode-select --install` in order to use the git+<package> command.
